pkg/antctl/commandoption.go

```diff
@@ -0,0 +1,350 @@
+// Copyright 2019 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package antctl
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/fatih/structtag"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+
+	"github.com/vmware-tanzu/antrea/pkg/antctl/handlers"
+)
+
+// commandGroup is used to group commands, it could be specified in CommandOption.
+// The default commandGroup of a CommandOption is `flat` which means the command generated by the CommandOption should
+// is a direct sub-command of the root command. For any other commandGroup, the antctl framework will generate a
+// sub-command direct belongs to the root group, any command generated by these CommandOptions will need to be invoked
+// like:
+//   antctl <commandGroup> <command>
+type commandGroup uint
+
+const (
+	flat commandGroup = iota
+	get
+)
+
+var groups = map[commandGroup]*cobra.Command{}
+
+// String returns the string representation of the command group.
+func (g commandGroup) String() string {
+	switch g {
+	case flat:
+		return "flat"
+	case get:
+		return "get"
+	default:
+		return "unknown"
+	}
+}
+
+// Command returns the dummy command of the command group.
+func (g commandGroup) Command(root *cobra.Command) (cmd *cobra.Command) {
+	switch g {
+	case get:
+		if cmd, ok := groups[get]; ok {
+			return cmd
+		}
+		cmd = &cobra.Command{
+			Use:   g.String(),
+			Short: "Get the status or resource of a topic",
+			Long:  "Get the status or resource of a topic",
+		}
+		root.AddCommand(cmd)
+		groups[get] = cmd
+	case flat:
+		cmd = root
+	default:
+		cmd = root
+	}
+	return cmd
+}
+
+const (
+	// TagKey is the tag name of the antctl specific annotation.
+	// For example:
+	// 	type FooResponse struct {
+	// 		Bar BarType `antctl:"key"`
+	// 	}
+	// If the field is annotated with antctl:"key", the framework assumes this field could be used to retrieve a unique
+	// Response, thus the framework will generate corresponding arg options to the cobra.Command.
+	TagKey = "antctl"
+	// TagOptionKeyArg is the option for antctl annotation. It tells the antctl the field is a primary key.
+	TagOptionKeyArg = "key"
+)
+
+// ArgOption defines a argument to be used in a RequestOption.
+type ArgOption struct {
+	Name      string
+	FieldName string
+	Usage     string
+	Key       bool
+}
+
+// CommandOption defines options to create a cobra.Command for antctl.
+type CommandOption struct {
```

Comment:
I find the name `CommandOption` confusing. I think it should be one of `Command`, `CommandOptions` or `CommandDefinition`. My favorite is the last one.

Reply:
OK, `CommandDefinition `.